### PR TITLE
Fix/ friend request logic and real time feedback for friend requests

### LINKED ISSUE #267: Fix the friend request logic and update user's friend request status in real time

### LINKED ISSUE #275: Make notifications for friend requests in profile screen update in real time

### PR DESCRIPTION
**Pull request by BogdanCDev**
## Summary

This PR adapts the friend request logic in the UserViewModel to ensure users can now see real-time updates to the state of their friend requests and notification count. It also introduces a new method, clearFriendRequestState, to handle the deletion of lingering friend request state when necessary. Additionally, it implements the listenToFriendRequests method to enable live updates for incoming friend requests and notification counts.

A new "Accept" button has been added to the user contact buttons in both search results and user profile views, allowing users to directly accept friend requests in context. These changes enhance the responsiveness, maintainability, and usability of the friend request feature.

This is a bug fix and a feature improvement.

## Changes

- **Logic/Controllers:** 
  - Added the clearFriendRequestState method in UserViewModel to remove lingering friend request states both in the repository and local state flows (sentFriendRequests and friendRequests).
  Refactored acceptFriendRequest, removeContact, and deleteFriendRequest methods to use clearFriendRequestState for improved consistency.
  - Introduced the listenToFriendRequests method in UserViewModel to set up a Firestore listener for real-time updates to friend requests.
  - Updated _notificationCount in UserViewModel to reflect changes in friend requests in real-time.
  - Removed redundant manual fetch calls for friend requests and notification count.
  - Improved error handling and logging for debugging state changes.
  - Added support for a dynamic "Accept" button:
    - Refactored UserSearchResultItem to include an "Accept" option for friend requests in the search results.
    - Refactored UserProfileContent to include the same "Accept" functionality for friend requests in the user profile view.
    - Updated the button color and action to align with user state (e.g., "Add", "Cancel", "Remove", or "Accept").

- **Repositories:**
  - Updated FriendRequestRepository interactions to ensure consistent state removal using the new clearFriendRequestState logic.
   - Added support for listening to real-time friend request updates in FriendRequestRepository.

- **Bug Fixes/Improvements:** 
  - Fixed the issue where the state of friend requests (e.g., "Cancel" or "Add" button) was not updating correctly in some cases.
  - Ensured notification badges update dynamically in the bottom navigation menu without requiring a manual refresh.
  - Improved the reactivity of the friend request button states in both the search and profile views.
  - Added `Log.d` statements for better tracking of state changes.

## Checklist

Ensure all / most of these are checked before the pull request is submitted.
- [X] This PR resolves #267 , #275 
- [X] Tests have been added for new functionality.
- [X] Documentation has been updated
- [X] Added screenshots

##  Testing

Explain how this feature was tested and what kind of testing (unit, integration, UI) has been done. Mention platforms or configurations used for testing.
- **Manual Testing:**
    - [X] Tested on emulator / physical device.
- **Unit Tests**
    - [X] Added units tests for this

##  Related Issues/Tickets

Closes #267
Closes #275 

## Screenshots

https://github.com/user-attachments/assets/98823a11-02b3-49cb-b81a-53ad7dfafbf6

https://github.com/user-attachments/assets/4aee12f9-4a65-4591-b25e-d8b2128e8b66

https://github.com/user-attachments/assets/a7218479-2e95-4699-832e-54e5da4b415d

### CODE REVIEW

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 1]
```diff
@@ -193,45 +206,111 @@ open class UserViewModel(
...
-    deleteFriendRequest(friendRequestId)
+    val contacts = user.value?.contacts?.toMutableSet()
+    val newUser = user.value!!.copy()
+    contacts?.add(userId)
+    newUser.contacts = contacts?.toList().orEmpty()
+    if (user.value != null) {
+      updateUser(newUser)
+      // Deletes Friend Request since the user has been added as a contact
+      deleteFriendRequest(friendRequestId)
+    }
```
Comment (ioanajianu) [inline app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:225]: `if (user.value != null)` is redundant after `val newUser = user.value!!.copy()`. Can we instead do something like     `val currentUser = user.value ?: return` at the beginning of the function?

File: app/src/main/java/com/android/voyageur/ui/profile/ProfileScreen.kt
Comment (ioanajianu) [inline app/src/main/java/com/android/voyageur/ui/profile/ProfileScreen.kt:1]: I have a few comments to make regarding the behavior of the app. I don't have the whole context of #266 so maybe these changes belong there and not on your PR, but I'll write them here as well:
- If I’m in the search screen and I click the “Remove” button then the button state changes to “Cancel”. Shouldn’t it be changed to “Add” directly?
- I tested this feature with Mircea's device. If Mircea accepts the friend request I sent him, a red dot still appears on the profile bottom bar as if he has 1 pending friend request. Is it possible to make the number of friend requests and this UI feature synchronized?
- if Mircea has a friend request from me, he can still click on add (Send me a friend request as well). If he sends one, and I accept it, then he has me among his contacts but I also appear on his friend requests list. What do you think about the button turning to something like “Accept” when you have a friend request from that person?

Let me know if these are planned to be addressed in the other PR or you think some bugs can be addressed in this one as well.

**Review by ioanajianu — COMMENTED** [review]
Hi Bogdan, thanks for implementing this fix! I tested this feature and now indeed both users get their contact list updated when one accepts the friend request or removes a friend.
I have a few things to add:
- could you please add a few simple suggestive sentences to the PR description summary? Something like we discussed offline, what was the behavior previously, and what is the behavior with your commits. The current summary uses some terms for which you need context, and is a bit ambiguous (before asking you offline, I wasn't sure what "real-time updates" refers to in this context).
- Documentation: Thanks for adding documentation to the code! The method descriptions and inline comments help with understanding. However, in the future could you please try to add the documentation before assigning reviewers? This way reviewers spend less time trying to understand the code. No worries though :)
- Testing: Thanks for adding tests! you currently have a coverage of 70.9%. I know that, given the nature of the UserViewModel and your changes, that's not easy to obtain. Since we aim for a 80% coverage on all PRs, could you please explain in the PR description why a higher coverage is not possible? Or, I see that right now the failure cases aren't covered. Is there a way to mock those and check that the logs at least are displayed?

Overall, this is an important fix that isn't easy to implement, and the overall code structure looks good. Great work!

**Comment by g0g05arui** [comment]
@BogdanCDev As I see our tasks were really similar and you need the listeners in your PR as well for your fix to work properly  I suggest you take a look at my PR(https://github.com/swent-t16/Voyageur/pull/255) which is in a draft state and you just take (if it's helpful) the code for listeners (it short, ~ 50 / 60 lines) as it doesn't make sense that the both of us work on the same thing. Maybe we should have clarified this situation before starting to work / when we created the tasks for this sprint, but at this moment I don't have the bandwidth to finish the PR in time without blocking this.

**Comment by ioanajianu** [comment]
As this PR is getting quite large, @g0g05arui could you have a look as well? I think you have the most context.

**Comment by g0g05arui** [comment]
Sure, I will take a look now

File: app/src/main/java/com/android/voyageur/ui/search/SearchUserProfile.kt [hunk 2]
```diff
@@ -84,28 +84,31 @@ fun SearchUserProfileScreen(userViewModel: UserViewModel, navigationActions: Nav
...
       onAddContact = { userViewModel.sendContactRequest(userData.id) },
       onRemoveContact = { userViewModel.removeContact(userData.id) },
       onCancelRequest = {
         val requestId = userViewModel.getSentRequestId(userData.id)
         if (requestId != null) {
           userViewModel.deleteFriendRequest(requestId)
         }
+      },
+      onAcceptRequest = {
+        if (friendRequest != null) {
```
Comment (g0g05arui) [inline app/src/main/java/com/android/voyageur/ui/search/SearchUserProfile.kt:110]: Why are we null checking here? If the friendRequest is null there shouldn't be an accept button right?

File: app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt [hunk 3]
```diff
@@ -418,19 +417,14 @@ fun UserSearchResultItem(
...
+  val isContactAdded = currentUser?.contacts?.contains(user.id) ?: false
+  val isRequestPending = sentFriendRequests.any { it.to == user.id }
+  val isRequestReceived = receivedFriendRequests.any { it.from == user.id }
 
-  val isRequestPending by
-      remember(sentFriendRequests, user) {
-        derivedStateOf { sentFriendRequests.any { it.to == user.id } }
-      }
-  LaunchedEffect(isContactAdded, isRequestPending) {
+  LaunchedEffect(isContactAdded, isRequestPending, isRequestReceived) {
```
Comment (g0g05arui) [inline app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt:427]: before merging you should remove this log which I assume was used for debugging purposes

File: app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt [hunk 4]
```diff
@@ -475,50 +464,69 @@ fun UserSearchResultItem(
...
                     }
+                  },
+                  enabled = isConnected,
+                  colors =
+                      ButtonDefaults.buttonColors(
+                          containerColor = MaterialTheme.colorScheme.tertiary),
+                  shape = RoundedCornerShape(20.dp),
+                  modifier = Modifier.width(100.dp).height(40.dp)) {
+                    Text(
+                        text = "Accept",
```
Comment (ioanajianu) [inline app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt:487]: Could you please use the string resources (`R.string`) for "Accept"?

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 5]
```diff
@@ -88,21 +102,31 @@ open class UserViewModel(
...
           // User is signed in, start listening to user data
           loadUser(firebaseUser.uid, firebaseUser)
-          // Fetch sent friend requests
+          // Listen to friend requests (incoming)
+          listenToFriendRequests()
+          // Listen to friend requests (sent)
           getSentFriendRequests()
         } else {
-          // User is signed out, clear user data and remove listeners
+          // User is signed out, clear data and remove listeners
```
Comment (g0g05arui) [inline app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:110]: Good catch to check the sign out behaviour

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 6]
```diff
@@ -88,21 +102,31 @@ open class UserViewModel(
...
           userListenerRegistration?.remove()
           userListenerRegistration = null
           _user.value = null
           _isLoading.value = false
+          sentFriendRequestsListener?.remove()
```
Comment (g0g05arui) [inline app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:115]: I don't understand the need for the sentFriendRequestListener. I am operating under the assumption that the user is only logged in from one device at a time, so the only way he can send a friend request is from the add so maybe there is no use for this.

**Review by g0g05arui — COMMENTED** [review]
While I appreciate the fact that you added a lot of tests, I believe that adding over 900 lines to a single test file may be a bit excessive. It would make code a bit more concise if some tests were combined, which I think they could be.

This feature is great and will help with the notifications system, great job!

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 1 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:243]: Thank you for the suggestion! You're absolutely right—if (user.value != null) is redundant after val newUser = user.value!!.copy(), as the !! operator already guarantees that user.value is not null. Using val currentUser = user.value ?: return at the beginning of the function simplifies the code and avoids repetitive null checks.

File: app/src/main/java/com/android/voyageur/ui/search/SearchUserProfile.kt (hunk 2 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/search/SearchUserProfile.kt:110]: Thank you for the feedback! While I understand your point that the Accept button should not be displayed if friendRequest is null, I believe the null check here is still necessary for robustness. 
Firstly, the UI logic (whether the button is displayed) and the backend logic (processing the request) should remain decoupled. This allows for more flexible and maintainable code. Even if the UI is correct now, there could be scenarios in the future where this assumption doesn't hold. 
Secondly, including the null check ensures the app doesn't run into unexpected crashes or undefined behaviour in edge cases where the friendRequest might be null due to synchronisation issues, a race condition, or a state update after the button was displayed.

File: app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt (hunk 3 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt:427]: Thank you for pointing that out! The log statement in the LaunchedEffect was indeed intended for debugging purposes to monitor recompositions during development. You’re absolutely right that it should be removed before merging to ensure cleaner production code.

Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt:427]: Thank you for pointing that out! The log statement in the LaunchedEffect was indeed intended for debugging purposes to monitor recompositions during development. You’re absolutely right that it should be removed before merging to ensure cleaner production code.

File: app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt (hunk 4 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt:487]: Thank you for pointing that out! You're absolutely right—using string resources (R.string) for "Accept" ensures better maintainability and supports localization. I'll update the code to use stringResource(R.string.accept) for consistency. Appreciate your feedback!

File: app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt [hunk 7]
```diff
@@ -475,50 +464,69 @@ fun UserSearchResultItem(
...
+                                isRequestPending -> MaterialTheme.colorScheme.secondary
+                                else -> MaterialTheme.colorScheme.primary
+                              }),
+                  shape = RoundedCornerShape(20.dp),
+                  modifier = Modifier.width(100.dp).height(40.dp)) {
+                    Text(
+                        text =
+                            when {
+                              isContactAdded -> "Remove"
+                              isRequestPending -> "Cancel"
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/search/SearchScreen.kt:521]: Done!

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 5 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:110]: Thank you!

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt (hunk 6 above)
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:115]: Thank you for the feedback! Your assumption about single-device usage is valid in certain scenarios, but the sentFriendRequestListener serves a broader purpose. Let me clarify:
Although we may assume single-device usage, Firebase allows users to log in on multiple devices simultaneously. If a friend request is sent from another device, the current device would be unaware of this action without real-time updates. The listener ensures that such actions are reflected immediately.
Also, even if we maintain the assumption of single-device usage, the listener provides real-time updates to the UI. This eliminates the need to fetch the sent requests list repeatedly, improving user experience and reducing server load.
Even more, there is the issue of edge cases like: Network issues that might cause delayed requests or retries, and without the listener, these might not be reflected until a manual refresh. 
Removing this listener might simplify the code but at the cost of user experience and potential desynchronization across devices or sessions.

**Comment by BogdanCDev** [comment]
> While I appreciate the fact that you added a lot of tests, I believe that adding over 900 lines to a single test file may be a bit excessive. It would make code a bit more concise if some tests were combined, which I think they could be.
> 
> This feature is great and will help with the notifications system, great job!

Thank you for your feedback! Regarding the length of the test file, I intentionally added extensive test cases to ensure full coverage of all edge cases and scenarios for this feature. While I understand the concern about file size, I believe having each test as a separate case improves clarity and maintainability. Combining tests could introduce coupling between scenarios, making it harder to identify the root cause of failures in the future.

**Comment by BogdanCDev** [comment]
> Hi Bogdan, thanks for implementing this fix! I tested this feature and now indeed both users get their contact list updated when one accepts the friend request or removes a friend. I have a few things to add:
> 
> * could you please add a few simple suggestive sentences to the PR description summary? Something like we discussed offline, what was the behavior previously, and what is the behavior with your commits. The current summary uses some terms for which you need context, and is a bit ambiguous (before asking you offline, I wasn't sure what "real-time updates" refers to in this context).
> * Documentation: Thanks for adding documentation to the code! The method descriptions and inline comments help with understanding. However, in the future could you please try to add the documentation before assigning reviewers? This way reviewers spend less time trying to understand the code. No worries though :)
> * Testing: Thanks for adding tests! you currently have a coverage of 70.9%. I know that, given the nature of the UserViewModel and your changes, that's not easy to obtain. Since we aim for a 80% coverage on all PRs, could you please explain in the PR description why a higher coverage is not possible? Or, I see that right now the failure cases aren't covered. Is there a way to mock those and check that the logs at least are displayed?
> 
> Overall, this is an important fix that isn't easy to implement, and the overall code structure looks good. Great work!

Thank you for taking the time to test the feature and provide such detailed feedback—I really appreciate it!

- PR Description: Thanks for pointing out the ambiguity in the summary. I've updated the PR description to include clearer and more suggestive sentences about the previous and current behavior, as we discussed offline. This should make it more accessible for anyone reviewing it in the future.
- Documentation: I completely understand your point about having documentation ready before assigning reviewers. I'll make sure to prioritize adding documentation earlier in the process for future PRs to save reviewers' time. Thank you for the feedback and for being patient with this one.
- Testing: I appreciate your suggestion regarding coverage. I’ve implemented tests for the failure cases to ensure those scenarios are handled, and I'll make sure the final coverage has the 80% line coverage.

Thanks again for your review and for highlighting areas for improvement!

File: app/src/main/java/com/android/voyageur/ui/profile/ProfileScreen.kt
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/profile/ProfileScreen.kt:1]: > I have a few comments to make regarding the behavior of the app. I don't have the whole context of #266 so maybe these changes belong there and not on your PR, but I'll write them here as well:
> 
> * If I’m in the search screen and I click the “Remove” button then the button state changes to “Cancel”. Shouldn’t it be changed to “Add” directly?
> * I tested this feature with Mircea's device. If Mircea accepts the friend request I sent him, a red dot still appears on the profile bottom bar as if he has 1 pending friend request. Is it possible to make the number of friend requests and this UI feature synchronized?
> * if Mircea has a friend request from me, he can still click on add (Send me a friend request as well). If he sends one, and I accept it, then he has me among his contacts but I also appear on his friend requests list. What do you think about the button turning to something like “Accept” when you have a friend request from that person?
> 
> Let me know if these are planned to be addressed in the other PR or you think some bugs can be addressed in this one as well.

- Button State on “Remove”: I’ve updated the behavior so that when the "Remove" button is clicked, it changes directly to “Add” instead of “Cancel.” This should make the interaction more intuitive.
- Red Dot on Profile Bottom Bar: I’ve synchronized the number of friend requests with the UI, so the red dot now updates correctly when a friend request is accepted. This issue should be resolved.
- Add Button Behavior with Existing Friend Request: I’ve implemented the suggested behavior where the button changes to “Accept” if there’s already a friend request from that person. This ensures the flow is more consistent and avoids duplicate requests.

All these changes have been addressed in this PR. If you notice anything else during testing, feel free to let me know. Thanks again for the great suggestions!

**Comment by BogdanCDev** [comment]
> @BogdanCDev As I see our tasks were really similar and you need the listeners in your PR as well for your fix to work properly I suggest you take a look at my PR(#255) which is in a draft state and you just take (if it's helpful) the code for listeners (it short, ~ 50 / 60 lines) as it doesn't make sense that the both of us work on the same thing. Maybe we should have clarified this situation before starting to work / when we created the tasks for this sprint, but at this moment I don't have the bandwidth to finish the PR in time without blocking this.
> 
> I will close / delete the issue I was assigned and close the PR.

Thank you for bringing this up and for the opportunity to use the listener code from your draft PR (#255). I’ll review it and incorporate the relevant parts into my PR to ensure the fix works as expected.

You’re absolutely right that we could have clarified this earlier, but I appreciate your proactive approach to avoid blocking the progress of the sprint. If there’s anything else related to this or where I can assist, feel free to let me know.

**Comment by BogdanCDev** [comment]
> @BogdanCDev As I see our tasks were really similar and you need the listeners in your PR as well for your fix to work properly I suggest you take a look at my PR(#255) which is in a draft state and you just take (if it's helpful) the code for listeners (it short, ~ 50 / 60 lines) as it doesn't make sense that the both of us work on the same thing. Maybe we should have clarified this situation before starting to work / when we created the tasks for this sprint, but at this moment I don't have the bandwidth to finish the PR in time without blocking this.
> 
> I will close / delete the issue I was assigned and close the PR.

Thank you for bringing this up and for the opportunity to use the listener code from your draft PR (#255). I've reviewed it and I have incorporated the relevant parts into my PR to ensure the fix works as expected.

You’re absolutely right that we could have clarified this earlier. If there’s anything else related to this or where I can assist, feel free to let me know.

**Review by ioanajianu — APPROVED** [review]
Thanks Bogdan for fixing the bugs I suggested earlier and for implementing the code changes. 

The code coverage is greater than 80%, nice work! While I agree with Mihai that one of the test files is big, I think that for tests it’s more important, as you said, to test one functionality per test so that it is easier to debug. 
However, I think there was an issue here that the changes themselves were quite large.. I know that this wasn’t intended and there was a misunderstanding when trying to solve #267 and #275 separately, and this PR ended up solving both. But, in the future, please try to keep PRs small, so that it’s easier for reviewers to catch errors, and thus the code quality is higher.
One quick suggestion: it’s best to mark comments as resolved only after pushing the corresponding changes. This keeps things clearer for reviewers.

Overall, this was a challenging fix, and the code looks good. Congrats Bogdan, fantastic work!

**Comment by BogdanCDev** [comment]
> Thanks Bogdan for fixing the bugs I suggested earlier and for implementing the code changes.
> 
> The code coverage is greater than 80%, nice work! While I agree with Mihai that one of the test files is big, I think that for tests it’s more important, as you said, to test one functionality per test so that it is easier to debug. However, I think there was an issue here that the changes themselves were quite large.. I know that this wasn’t intended and there was a misunderstanding when trying to solve #267 and #275 separately, and this PR ended up solving both. But, in the future, please try to keep PRs small, so that it’s easier for reviewers to catch errors, and thus the code quality is higher. One quick suggestion: it’s best to mark comments as resolved only after pushing the corresponding changes. This keeps things clearer for reviewers.
> 
> Overall, this was a challenging fix, and the code looks good. Congrats Bogdan, fantastic work!

Thank you for the detailed feedback and kind words! Regarding issue #275, it was added based on your suggestions, and since it aligned with the changes here, it seemed appropriate to include it in this PR. Thank you for suggesting it—it helped ensure we covered all related aspects effectively.
